internal/proxy/util.go

```diff
@@ -593,3 +603,114 @@ func parseGuaranteeTs(ts, tMax typeutil.Timestamp) typeutil.Timestamp {
 	}
 	return ts
 }
+
+func validateName(entity string, nameType string) error {
+	entity = strings.TrimSpace(entity)
+
+	if entity == "" {
+		return fmt.Errorf("%s should not be empty", nameType)
+	}
+
+	invalidMsg := fmt.Sprintf("invalid %s: %s. ", nameType, entity)
+	if int64(len(entity)) > Params.ProxyCfg.MaxNameLength {
+		msg := invalidMsg + fmt.Sprintf("the length of %s must be less than ", nameType) +
+			strconv.FormatInt(Params.ProxyCfg.MaxNameLength, 10) + " characters."
+		return errors.New(msg)
+	}
+
+	firstChar := entity[0]
+	if firstChar != '_' && !isAlpha(firstChar) {
+		msg := invalidMsg + fmt.Sprintf("the first character of %s must be an underscore or letter.", nameType)
+		return errors.New(msg)
+	}
+
+	for i := 1; i < len(entity); i++ {
+		c := entity[i]
+		if c != '_' && c != '$' && !isAlpha(c) && !isNumber(c) {
+			msg := invalidMsg + fmt.Sprintf("%s can only contain numbers, letters, dollars and underscores.", nameType)
+			return errors.New(msg)
+		}
+	}
+	return nil
+}
+
+func ValidateRoleName(entity string) error {
+	return validateName(entity, "role name")
+}
+
+func ValidateObjectName(entity string) error {
+	entity = strings.TrimSpace(entity)
+	if entity == "" {
+		return fmt.Errorf("objectName should not be empty")
+	}
+
+	if int64(len(entity)) > Params.ProxyCfg.MaxNameLength {
+		msg := fmt.Sprintf("invalid object name: %s. The length of resource name must be less than %s characters",
+			entity, strconv.FormatInt(Params.ProxyCfg.MaxNameLength, 10))
+		return errors.New(msg)
+	}
+	return nil
+}
+
+func ValidateObjectType(entity string) error {
+	return validateName(entity, "ObjectType")
+}
+
+func ValidatePrincipalName(entity string) error {
+	return validateName(entity, "PrincipalName")
+}
+
+func ValidatePrincipalType(entity string) error {
+	return validateName(entity, "PrincipalType")
+}
+
+func ValidatePrivilege(entity string) error {
+	return validateName(entity, "Privilege")
+}
+
+func GetCurUserFromContext(ctx context.Context) (string, error) {
+	md, ok := metadata.FromIncomingContext(ctx)
+	if !ok {
+		return "", fmt.Errorf("fail to get md from the context")
+	}
+	authorization := md[strings.ToLower(util.HeaderAuthorize)]
+	if len(authorization) < 1 {
+		return "", fmt.Errorf("fail to get authorization from the md")
+	}
+	token := authorization[0]
+	rawToken, err := crypto.Base64Decode(token)
+	if err != nil {
+		return "", fmt.Errorf("fail to decode the token, token: %s", token)
+	}
+	secrets := strings.SplitN(rawToken, util.CredentialSeperator, 2)
+	if len(secrets) < 2 {
+		return "", fmt.Errorf("fail to get user info from the raw token, raw token: %s", rawToken)
+	}
+	username := secrets[0]
```

Comment:
is it one element at least within the secrets array?

Reply:
yes, including username and password.